[DataPipe] Improving debug message when argument is a tuple/list of DataPipes

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76134

The previous PR #75618 neglected to handle the case where the argument is a tuple/list of DataPipes, as it is the case for `zip`, `mux`, etc. This PR addresses that issue.

Now the message looks like the following, instead of a bunch of `<...>` inside the tuple.
```
This exception is thrown by __iter__ of ZipperIterDataPipe(datapipes=(IterableWrapperIterDataPipe, IterableWrapperIterDataPipe, IterableWrapperIterDataPipe))
```

NivekT committed:
[DataPipe] Improving debug message when argument is a tuple/list of DataPipes

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -348,6 +348,9 @@ def _simplify_obj_name(obj) -> str:
             return str(obj.__class__.__qualname__)
         except Exception as _:
             return default_str
+    elif isinstance(obj, (list, tuple)):
+        inner = ', '.join([_simplify_obj_name(ele) for ele in obj])
+        return f"[{inner}]" if isinstance(obj, list) else f"({inner})"
```
This can be endless data structure than can contain `DataPipe` (`dict`, `set`, all custom container objects for the custom DataPipe). Why don't we simply add a `__str__` function to `IterDataPipe` class? Then we can directly use the output from `str(obj)`.

NivekT replied:
Yea, I think overwriting `__repr__` and `__str__` is the way to go. I checked the exception message it works.

NivekT committed:
Update on "[DataPipe] Improving debug message when argument is a tuple/list of DataPipes"


The previous PR #75618 neglected to handle the case where the argument is a tuple/list of DataPipes, as it is the case for `zip`, `mux`, etc. This PR addresses that issue.

Now the message looks like the following, instead of a bunch of `<...>` inside the tuple.
```
This exception is thrown by __iter__ of ZipperIterDataPipe(datapipes=(IterableWrapperIterDataPipe, IterableWrapperIterDataPipe, IterableWrapperIterDataPipe))
```


[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -132,6 +132,14 @@ def set_reduce_ex_hook(cls, hook_fn):
             raise Exception("Attempt to override existing reduce_ex_hook")
         IterDataPipe.reduce_ex_hook = hook_fn
 
+    def __repr__(self):
+        # Instead of showing <torch. ... .MapperIterDataPipe object at 0x.....>, return the class name
+        return str(self.__class__.__qualname__)
+
+    def __str__(self):
+        # Instead of showing <torch. ... .MapperIterDataPipe object at 0x.....>, return the class name
+        return str(self.__class__.__qualname__)
+
```
What's the difference between `__str__` and `__repr__`?
And, can we add a registration stub in it? Then, when we work on the graph visualization or buffer/spec visualization, we will be able to override these functions.

NivekT replied:
Certain built-in functions choose to call `__repr__` (e.g. when exception is raised, `__repr__(obj)` is called). 

What do you mean by registration stub? Something similar to the `set_state_hook`?

NivekT replied:
The stubs have been added!

NivekT committed:
Update on "[DataPipe] Improving debug message when argument is a tuple/list of DataPipes"


The previous PR #75618 neglected to handle the case where the argument is a tuple/list of DataPipes, as it is the case for `zip`, `mux`, etc. This PR addresses that issue.

Now the message looks like the following, instead of a bunch of `<...>` inside the tuple.
```
This exception is thrown by __iter__ of ZipperIterDataPipe(datapipes=(IterableWrapperIterDataPipe, IterableWrapperIterDataPipe, IterableWrapperIterDataPipe))
```


[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Improving debug message when argument is a tuple/list of DataPipes"


The previous PR #75618 neglected to handle the case where the argument is a tuple/list of DataPipes, as it is the case for `zip`, `mux`, etc. This PR addresses that issue.

Now the message looks like the following, instead of a bunch of `<...>` inside the tuple.
```
This exception is thrown by __iter__ of ZipperIterDataPipe(datapipes=(IterableWrapperIterDataPipe, IterableWrapperIterDataPipe, IterableWrapperIterDataPipe))
```


[ghstack-poisoned]

ejguan reviewed (approved):
LGTM. Thank you

NivekT committed:
Update on "[DataPipe] Improving debug message when argument is a tuple/list of DataPipes"


The previous PR #75618 neglected to handle the case where the argument is a tuple/list of DataPipes, as it is the case for `zip`, `mux`, etc. This PR addresses that issue.

Now the message looks like the following, instead of a bunch of `<...>` inside the tuple.
```
This exception is thrown by __iter__ of ZipperIterDataPipe(datapipes=(IterableWrapperIterDataPipe, IterableWrapperIterDataPipe, IterableWrapperIterDataPipe))
```


[ghstack-poisoned]

NivekT commented:
@pytorchbot merge this please
